docs(oss): models (#249)
langchain-ai/docs: src/oss/langchain-models.mdx

If you want to receive the output for each individual input as it is finishes
generating, you can stream results with `batch_as_completed()`:

```python Yield responses upon completion
for response in model.batch_as_completed([
    "Why do parrots have colorful feathers?",
    "How do airplanes fly?",
    "What is quantum computing?"
]):
    print(response)
```
<Note>
    When using `batch_as_completed()`, results may arrive out of order. Each
    includes the input index for matching to reconstruct the original order if
    needed.
</Note>

<Tip>
    When processing a large number of inputs using `batch()` or
    `batch_as_completed()`, you may want to control the maximum number of
    parallel calls. This can be done by setting the `max_concurrency` attribute
    in the `RunnableConfig` dictionary.
    
    ```python Batch with max concurrency
    model.batch(
        list_of_inputs,
        config={
            'max_concurrency': 5,  # Limit to 5 parallel calls
        }
    )
    ```

    See the [`RunnableConfig`]() reference for a full list of supported
    attributes.
</Tip>

For more details on batching, see the [reference]().

> ## Context windows

Reviewer: short term memory section!

    interchangeably with `tool calling`.
</Note>

To make tools that you have defined available for use by a model, you must bind
them using `.bind_tools()`. In subsequent invocations, the model can choose to
call any of the bound tools as needed.

Let's define some example tools that we'll use throughout this section:

<ExampleTools />

Some model providers offer built-in tools that can be enabled via model
parameters. In such cases, you do not need to define or bind those tools to your
model. Check the respective [provider reference]() for details.

<CodeGroup>
    ```python Binding user tools highlight={7}
    from pydantic import BaseModel, Field

    class GetWeather(BaseModel):
        """Get the current weather in a given location"""
        location: str = Field(..., description="The city and state, e.g. Boston, MA")

    model_with_tools = model.bind_tools([GetWeather])

    response = model_with_tools.invoke("What's the weather like in Boston?")
    for tool_call in response.tool_calls:
        # View tool calls made by the model
        print(f"Tool: {tool_call['name']}")
        print(f"Args: {tool_call['args']}")
    ```
    

>     ```python Built-in tool usage with citations

Reviewer 2: Because binding these tools are not standardized at this point, I would just include a `<Tip>` mentioning built-in tools, such as server-side web search and code interpreters, and directing readers to the Integrations tab to read about their provider of choice.

    response = model.invoke("Search the web for the latest news on AI.")
    content = response.content_blocks

    # content: [
    #     TextContentBlock(
    #         text = "The latest news articles on AI..."),
    #         annotations = [
    #             Citation(
    #                 title="AI breakthrough in natural language processing",
    #                 url="https://example.com/ai-news",
    #                 cited_text="Researchers have...",
    #         )]
    # ])
    ```
</CodeGroup>

{/* TODO a switcher to iterate through all different tool schema options */}
{/* TODO model switcher to show this example for our top 5 providers */}

When binding user-defined tools, the model's response includes a **request** to
execute a tool. It is up to you to perform the requested action and return the
result back to the model for use in subsequent reasoning.

<AccordionGroup>

>     <Accordion title="Tool execution loop">

Reviewer 2: Let's link out to langgraph sections so people don't think they need to orchestrate this themselves (especially with parallel tool calling section, below)

        When a model returns tool calls, you need to execute the tools and pass the
        results back to the model. This creates a conversation loop where the model
        can use tool results to generate its final response.

        ```python Tool execution loop
        from langchain_core.messages import HumanMessage

        # Bind tools to the model

>         model_with_tools = model.bind_tools([get_weather, get_time])

Reviewer 2: `get_time` is undefined at this point

        
        # Step 1: Model generates tool calls
        messages = [HumanMessage("What's the weather in Boston?")]
        ai_msg = model_with_tools.invoke(messages)
        messages.append(ai_msg)
        
        # Step 2: Execute tools and collect results
        for tool_call in ai_msg.tool_calls:
            # Execute the tool with the generated arguments

>             tool_result = get_weather.invoke(tool_call)

Reviewer 2: `get_weather` needs to be an instance of `BaseTool` at this point

            messages.append(tool_result)
        
        # Step 3: Pass results back to model for final response
        final_response = model_with_tools.invoke(messages)
        print(final_response.text)
        # "The current weather in Boston is 72°F and sunny."
        ```

        Each `ToolMessage` returned by the tool includes a `tool_call_id` that
        matches the original tool call, helping the model correlate results with
        requests.
    </Accordion>

    <Accordion title="Forcing tool calls">
        #### Forcing tool calls

        By default, the model has the freedom to choose which bound tool to use
        based on the user's input. However, you might want to force choosing a
        tool, ensuring the model uses either a particular tool or *any* tool
        from a given list:

        <CodeGroup>
        ```python Force use of specific tools
        model_with_tools = model.bind_tools([get_weather, calculate])
        
        response = model_with_tools.invoke(
            "What's the weather in Boston and calculate 15 + 25",

>             force_tool_calls=["get_weather", "calculate"]

Reviewer 2: is this hallucinated? :)

should be `tool_choice="<name of tool>"` or `tool_choice="any"` as inputs to `bind_tools`

        print(response)  # Movie(title="Inception", year=2010, director="Christopher Nolan", rating=8.8)
        ```
    </Tab>
    <Tab title="TypedDict">
        `TypedDict` provides a simpler alternative using Python's built-in
        typing, ideal when you don't need runtime validation.

        ```python
        from typing_extensions import TypedDict, Annotated

        class MovieDict(TypedDict):
            """A movie with details."""
            title: Annotated[str, ..., "The title of the movie"]
            year: Annotated[int, ..., "The year the movie was released"]
            director: Annotated[str, ..., "The director of the movie"]
            rating: Annotated[float, ..., "The movie's rating out of 10"]

        model_with_structure = model.with_structured_output(MovieDict)
        response = model_with_structure.invoke("Provide details about the movie Inception")
        print(response)  # {'title': 'Inception', 'year': 2010, 'director': 'Christopher Nolan', 'rating': 8.8}
        ```
    </Tab>
    <Tab title="JSON Schema">
        For maximum control or interoperability, you can provide a raw JSON

>         Schema. Some providers support `strict=True` for guaranteed schema

Reviewer 2: (nit) applies to TypedDict as well. but I believe this is just for OpenAI ([API ref](https://python.langchain.com/api_reference/openai/chat_models/langchain_openai.chat_models.base.ChatOpenAI.html#langchain_openai.chat_models.base.ChatOpenAI.with_structured_output))

                },
                "rating": {
                    "type": "number",
                    "description": "The movie's rating out of 10"
                }
            },
            "required": ["title", "year", "director", "rating"]
        }

        model_with_structure = model.with_structured_output(
            json_schema, 
            method="json_schema",
            strict=True  # Available on certain providers
        )
        response = model_with_structure.invoke("Provide details about the movie Inception")
        print(response)  # {'title': 'Inception', 'year': 2010, ...}
        ```
    </Tab>
</Tabs>

<Note>
    **Key considerations for structured outputs:**

    - **Method parameter**: Some providers support different methods
    (`'json_schema'`, `'function_calling'`, `'json_mode'`)
    - **Strict mode**: When available, `strict=True` guarantees the output will
    match your schema exactly
    - **Include raw**: Use `include_raw=True` to get both the parsed output and
    the raw AI message
    - **Validation**: Pydantic models provide automatic validation, while
    `TypedDict` and JSON Schema require manual validation
</Note>

> <Accordion title="Advanced usage with nested structures">

Reviewer 2: (nit) tabs for TypedDict?

    ```python
    from pydantic import BaseModel, Field
    from typing import List, Optional

    class Actor(BaseModel):
        name: str
        role: str

    class MovieDetails(BaseModel):
        title: str
        year: int
        cast: List[Actor]
        genres: List[str]
        budget: Optional[float] = Field(None, description="Budget in millions USD")

    model_with_structure = model.with_structured_output(
        MovieDetails, 

>         include_raw=True  # Returns both parsed and raw response

Reviewer 2: (nit) consider calling this out in a separate Accordion, instead of putting it in the "nested structures" one.

I do see that it's called out nicely in the `<Note>` above, but it could be useful to demonstrate. People often wonder how to get token counts and other metadata when using `with_structured_output`.

    )

    result = model_with_structure.invoke("Tell me about The Matrix movie")
    print(result["parsed"])  # Access the structured data
    print(result["raw"])     # Access the original AI message
    ```
</Accordion>

<Accordion title="Structured output with tool calling">
    You can combine structured outputs with tool calling to ensure tools return
    data in a specific format. This is particularly useful when tools need to

>     extract or process information in a standardized way.

Reviewer 2: This appears hallucinated.

    ```python
    from pydantic import BaseModel, Field
    from langchain_core.tools import tool

    class SearchResult(BaseModel):
        """A structured search result."""
        query: str = Field(description="The search query used")
        results: List[str] = Field(description="List of search results")
        confidence: float = Field(description="Confidence score 0-1")

    @tool(response_format=SearchResult)
    def web_search(query: str) -> SearchResult:
        """Search the web and return structured results."""
        # Your search logic here
        return SearchResult(
            query=query,
            results=["Result 1", "Result 2"],
            confidence=0.95
        )

    # Bind the tool to the model
    model_with_tools = model.bind_tools([web_search])

    # The tool will automatically return structured data
    response = model_with_tools.invoke("Search for information about quantum computing")
    if response.tool_calls:
        tool_result = response.tool_calls[0]
        print(tool_result["args"])  # Structured as SearchResult
    ```

    You can also use `with_structured_output()` alongside tools for extracting

>     structured information from tool responses:

Reviewer 2: This example does not make sense to me.

I think the tool bindings get clobbered?

    ```python
    class Analysis(BaseModel):
        """Analysis of search results."""
        summary: str = Field(description="Summary of findings")
        key_points: List[str] = Field(description="Key points extracted")
        needs_more_info: bool = Field(description="Whether more information is needed")

    # Create a model that uses tools and outputs structured data
    model_with_tools = model.bind_tools([web_search])
    structured_model = model_with_tools.with_structured_output(Analysis)

    # This will:
    # 1. Call the search tool if needed
    # 2. Process the results
    # 3. Return structured analysis
    result = structured_model.invoke(
        "Search for quantum computing and analyze the findings"
    )
    print(result)  # Analysis(summary="...", key_points=[...], needs_more_info=False)
    ```
</Accordion>

### Multimodality

Certain models can process and return non-textual data such as images, audio,
and video.

You can pass non-textual data to a model by creating [content blocks]() that
encapsulate the data:

<CodeGroup>    

>     ```python Image input

Reviewer 2: should we call out how to do inline base64 as well?

    from langchain.messages.content import create_image_block

    response = model.invoke([
        "Describe the content of this image.",
        create_image_block(url="https://example.com/path/to/image.jpg")

>     ])

Reviewer 2: IMO: strongly prefer
```python
response = model.invoke([
    {"type": "text", "text": "Describe the content of this image."},
    {"type": "image", "url": "https://example.com/path/to/image.jpg"},
])
```

    print(response)  # AIMessage("The image shows a beautiful sunset...")
    ```

    ```python Video input
    from langchain.messages.content import create_video_block

    video = create_video_block(
        base64="AAAAIGZ0eXBtcDQyAAAAAGlzb21tcDQyAAACAGlzb2...",
        mime_type="video/mp4"

>     )

Reviewer 2: don't think there are any providers for which this will work atm.

    response = model.invoke([
        "Describe the content of this video.",
        video
    ])
    print(response)  # AIMessage("The video shows a cat playing with a ball...")
    ```

    ```python Audio input
    from langchain.messages.content import create_audio_block

    response = model.invoke([
        "Describe the content of this audio clip.",

>         create_audio_block(file_id="file-abc123xyz456")

Reviewer 2: why file and not base64?

    ])
    print(response)  # AIMessage("The audio clip features a piano melody...")
    ```

    ```python File input (PDF)
    from langchain.messages.content import create_file_block

    file = create_file_block(
        base64="AAAAIGZ0eXBtcDQyAAAAAGlzb21tcDQyAAACAGlzb2...",
        mime_type="application/pdf",
        filename="document.pdf"
    )

    response = model.invoke([
        "Summarize the content of this PDF document.",
        file
    ])
    print(response)  # AIMessage("The document discusses the importance of...")
    ```
</CodeGroup>

On the other hand, models can return non-textual data as part of their response:

<CodeGroup>

>     ```python Image output

Reviewer 2: I would maybe just show images here and link to the Integrations page for users to learn more about their specific provider.

I think there are zero providers for which we support video or PDF, and one for which we support audio, but it's not a popular use case.

    response = model.invoke("Generate an image of a green parrot.")

>     image = response.content_blocks[0]

Reviewer 2: should traverse content here

reasoning entirely.

<CodeGroup>
    ```python Reasoning effort
    response = model.invoke(
        "Why do parrots have colorful feathers?",
        reasoning_effort="low"
    )
    print(response)
    ``` 

    ```python Disable reasoning
    response = model.invoke(
        "Why do parrots have colorful feathers?",
        reasoning_effort="none"
    )
    print(response)
    ```
</CodeGroup>

To see what reasoning capabilities are supported by a particular provider or
model, check the respective [reference]().

## Supported models

LangChain supports all major model providers, including OpenAI, Anthropic,
Google, Azure, AWS Bedrock, and more. Each provider offers a variety of models
with different capabilities. For a full list of supported models in LangChain,
see the [provider reference]().

## Advanced configuration

### Local models

LangChain supports running models locally on your own hardware. This is useful
for scenarios where data privacy is critical, or when you want to avoid the cost
of using a cloud-based model.

[Ollama]() is one of the easiest ways to run models locally. See the full list
of local integrations in the [reference]().

> ### Caching

Reviewer: how good do we feel about it? if not good - lets err on side of removing

    it fails due to issues like network timeouts or rate limits.
</ParamField>

<Info>
    To find all the parameters supported by a `ChatModel`, head to the
    [reference]().
</Info>

## Invocation

A `ChatModel` must be invoked to generate an output. There are three main
invocation methods, each suited to different use cases.

<Tip>
    Each invocation method has an
    <Tooltip tip="Asynchronous methods are useful for non-blocking operations, especially in web servers or applications that handle multiple tasks concurrently.">asynchronous</Tooltip>
    equivalent, typically prefixed with the letter `'a'`
    
    For example: `ainvoke()`, `astream()`, `abatch()`
    
    A full list of async methods can be found in the [reference]().
</Tip>

### Invoke

The most straightforward way to call a model is to use `invoke()` with a single
message or a list of messages.

A list of messages can be provided to a model to represent a conversation
history. Each message has a role to provide context as to who is speaking in the
sequence. See the [messages]() guide for more detail.
{/* TODO include three important topics in this callout */}

<CodeGroup>
```python Single message

> response = model.invoke("Why do parrots have colorful feathers?")

Reviewer: confusing that we dont use messages here or mention strings can be used instead

print(response)
```
<ConversationHistory/>
</CodeGroup>

### Stream

Most models can stream their output content while it is being generated. By
displaying output progressively, streaming significantly improves user
experience, particularly for longer responses.

Calling `stream()` returns an <Tooltip tip="An object that progressively provides access to each item of a collection, in order.">iterator</Tooltip>
that yields output chunks as they are produced. You can use a loop to process
each chunk in real-time:

```python Streaming
for chunk in model.stream("Why do parrots have colorful feathers?"):
    print(chunk)
```

> See the [streaming]() guide for more detail.

Reviewer: No - streaming guide covers agents. streaming llms should be here

    class GetWeather(BaseModel):
        """Get the current weather in a given location"""
        location: str = Field(..., description="The city and state, e.g. Boston, MA")

    model_with_tools = model.bind_tools([GetWeather])

    response = model_with_tools.invoke("What's the weather like in Boston?")
    for tool_call in response.tool_calls:
        # View tool calls made by the model
        print(f"Tool: {tool_call['name']}")
        print(f"Args: {tool_call['args']}")
    ```
    
    ```python Built-in tool usage with citations
    response = model.invoke("Search the web for the latest news on AI.")
    content = response.content_blocks

    # content: [
    #     TextContentBlock(
    #         text = "The latest news articles on AI..."),
    #         annotations = [
    #             Citation(
    #                 title="AI breakthrough in natural language processing",
    #                 url="https://example.com/ai-news",
    #                 cited_text="Researchers have...",
    #         )]
    # ])
    ```
</CodeGroup>

For more details on creating, using, and executing tools requests, see the
[tools]() guide.

{/* 

> If we want to include some tool info here, we can do so like this:

Reviewer: yes! all this stuff is good. things like defining tools will be in tools page, but other than that is largely about using tools in model, not agent. using tools in model should go here

    force_tool_calls=["GetWeather", "Calculator"]
)
```

```python Force use of any tool
response = model_with_tools.invoke(
    "What's the weather in both New York and London, and calculate 15 + 25",
    force_tool_calls="any"
)
```

</CodeGroup>

#### Parallel tool calls

TODO */}

### Structured outputs

Models can be requested to provide their response in a format matching a given
schema. This is useful for ensuring the output can be easily parsed and used in
subsequent processing.

```python Structured output highlight={12-14}
from pydantic import BaseModel, Field

class Movie(BaseModel):
    title: str = Field(..., description="The title of the movie")
    year: int = Field(..., description="The year the movie was released")
    director: str = Field(..., description="The director of the movie")
    rating: float = Field(..., description="The movie's rating out of 10")

model_with_structure = model.with_structured_output(Movie)
response = model_with_structure.invoke("Provide details about the movie Inception")
print(response)  # Movie(title="Inception", year=2010, director="Christopher Nolan", rating=8.8)
```

> See the [structured outputs]() guide for more info.

Reviewer: structured outputs is about structured outputs in agents, not models. any relevant info needs to go here

        reasoning_effort="low"
    )
    print(response)
    ``` 

    ```python Disable reasoning
    response = model.invoke(
        "Why do parrots have colorful feathers?",
        reasoning_effort="none"
    )
    print(response)
    ```
</CodeGroup>

To see what reasoning capabilities are supported by a particular provider or
model, check the respective [reference]().

## Supported models

LangChain supports all major model providers, including OpenAI, Anthropic,
Google, Azure, AWS Bedrock, and more. Each provider offers a variety of models
with different capabilities. For a full list of supported models in LangChain,
see the [provider reference]().

## Advanced configuration

### Local models

LangChain supports running models locally on your own hardware. This is useful
for scenarios where data privacy is critical, or when you want to avoid the cost
of using a cloud-based model.

[Ollama]() is one of the easiest ways to run models locally. See the full list
of local integrations in the [reference]().

### Caching

Chat model APIs can be slow and expensive to call. To help mitigate this,
LangChain provides an optional caching layer for `ChatModel` integrations. See

> the [caching]() guide for more details.

Reviewer: doesnt exist

### Rate limiting

Many chat model providers impose a limit on the number of invocations that can
be made in a given time period. If you hit a rate limit, you will typically
receive a rate limit error response from the provider, and will need to wait
before making more requests.

To help manage rate limits, `ChatModel` integrations accept a `rate_limiter`
parameter that can be provided during initialization to control the rate at

> which requests are made. See the [rate limiting]() guide for more details.

Reviewer: doesnt exist

Author: https://github.com/langchain-ai/langchain/blob/master/libs/core/langchain_core/rate_limiters.py
https://python.langchain.com/docs/how_to/chat_model_rate_limiting/

In addition to basic text generation, many models support the following
features:

* <Icon icon="hammer" size={16} /> [Tool calling](#tool-calling) - enables
models to interact with external services, APIs, and databases.
* <Icon icon="shapes" size={16} /> [Structured output](#structured-outputs) -
constrain the response format to match a provided schema.
* <Icon icon="image" size={16} /> [Multimodality](#multimodality) - the ability
to process data other than text, such as images.
* <Icon icon="brain" size={16} /> [Reasoning](#reasoning) - the ability to
perform multi-step reasoning to arrive at a conclusion.

## Basic usage

The primary interface for interacting with models in LangChain are the chat
model abstractions. They provide a consistent interface for accessing different
LLMs from various providers.

The easiest way start interacting with a model in LangChain is to use
`init_chat_model()` to initialize one from a model
[provider](/oss/integrations/providers) of your choice.

<CodeGroup>
```python Initialize a chat model
from langchain.chat_models import init_chat_model

model = init_chat_model("openai:gpt-5-nano")
response = model.invoke("Why do parrots talk?")

> ```

Reviewer 3: should we be calling out initializing direct from the provider package here also?

    model_provider="ollama",
)
response = model.invoke("Why do parrots talk?")
``` */}
</CodeGroup>

See [`init_chat_model()`](https://python.langchain.com/api_reference/langchain/chat_models/langchain.chat_models.base.init_chat_model.html)
for more detail.

### Key methods

<Card title="Invoke" href="#invoke" icon="paper-plane" arrow="true" horizontal>
    The model takes messages as input and returns messages after generating a
    full response.
</Card>
<Card title="Stream" href="#stream" icon="tower-broadcast" arrow="true" horizontal>
    Invoke the model, but stream the response as it is generated in real-time.
</Card>
<Card title="Batch" href="#batch" icon="folder-open" arrow="true" horizontal>
    Send multiple requests to a model in a batch for more efficient processing.
</Card>

<Info>
    In addition to chat models, LangChain provides support for other adjacent
    technologies, such as embedding models and vector stores. See the
    [integrations page](/oss/integrations/providers) for details.
</Info>

## Parameters

A chat model takes parameters that can be used to configure its behavior. The
full set of supported parameters vary by model and provider, but common ones

> include:

Reviewer 3: We should specify these are params for `init_chat_model`

Author: wdym? they work for both

Reviewer 3: to me I could see getting people confused on these being the only params you can use to create models for both init_chat_model and using the model initializer

            elif block["type"] == "tool_call_chunk":
                print(f"Tool call chunk: {block}")
            elif block["type"] == "text":
                print(block["text"])
            else:
                ...
    ```
</CodeGroup>

As opposed to `invoke()`, which returns a single `AIMessage` after the model has
finished generating its full response, `stream()` returns multiple
`AIMessageChunk` objects, each containing a portion of the output text.

Importantly, each `AIMessageChunk` in a stream is designed to be gathered into a
full message via summation:

```python Construct AIMessage
full = None  # None | AIMessageChunk
for chunk in model.stream("What color is the sky?"):
    full = chunk if full is None else full + chunk
    print(full.text)

# The
# The sky
# The sky is
# The sky is typically
# The sky is typically blue
# ...

print(full.content_blocks)
# [{"type": "text", "text": "The sky is typically blue..."}]
```

The resulting full message can be treated the same as a message that was
generated with [invoke](#invoke) - for example, it can be aggregated into a
message history and passed back to the model as conversational context.

<Warning>

>     Streaming is only effective if all steps in the program know how to process
>     an input stream; i.e., process an input chunk one at a time, and yield a
>     corresponding output chunk. In other words, streaming only works if the
>     entire application is designed to be streaming-capable. An application that
>     is **not** streaming-capable, for example, would be one that needs to store
>     the entire output in memory before it can be processed.

Reviewer 3: this is a little too wordy IMO

```suggestion
    Streaming only works if all steps in the program know how to process an
    stream of chunks. An application that isn't streaming-capable would be one
    that needs to store the entire output in memory before it can be processed.
```

</Warning>

<Accordion title='"Auto-Streaming" Chat Models'>
    <Note>
        This section requires some baseline knowledge of [callbacks]() and
        [events]().

>     </Note>

Reviewer 3: If we don't have links to these in v1 docs we shouldn't include them IMO

    to surface the chat model's output in real-time.
</Accordion>

<Accordion
    title="Stream events"
>    
    LangChain chat models can also stream semantic events using `astream_events()`.

    This simplifies filtering based on event types and other metadata, and will
    aggregate the full message in the background. See below for an example.

    ```python
    async for event in model.astream_events("Hello"):

        if event["event"] == "on_chat_model_start":
            print(f"Input: {event['data']['input']}")

        elif event["event"] == "on_chat_model_stream":
            print(f"Token: {event['data']['chunk'].text}")

        elif event["event"] == "on_chat_model_end":
            print(f"Full message: {event['data']['output'].text}")

        else:
            pass
    ```
    ```txt
    Input: Hello
    Token: Hi
    Token:  there
    Token: !
    Token:  How
    Token:  can
    Token:  I
    ...
    Full message: Hi there! How can I help today?
    ```
    
    See the [`astream_events()`](https://python.langchain.com/api_reference/langchain/agents/langchain.agents.agent.AgentExecutor.html#langchain.agents.agent.AgentExecutor.astream_events)

>     reference for event types and other details.

Author: do these work now?

        {"type": "text", "text": "Describe the content of this image."},
        {
            "type": "image",
            "base64": "AAAAIGZ0eXBtcDQyAAAAAGlzb21tcDQyAAACAGlzb2...",
            "mime_type": "image/jpeg",
        },
    ])

    # From provider-managed File ID
    response = model.invoke([
        {"type": "text", "text": "Describe the content of this image."},
        {"type": "image", "file_id": "file-abc123"},
    ])
    ```

    ```python PDF document input
    # From URL
    response = model.invoke([
        {"type": "text", "text": "Describe the content of this document."},
        {"type": "file", "url": "https://example.com/path/to/document.pdf"},
    ])

    # From base64 data
    response = model.invoke([
        {"type": "text", "text": "Describe the content of this document."},
        {
            "type": "file",
            "base64": "AAAAIGZ0eXBtcDQyAAAAAGlzb21tcDQyAAACAGlzb2...",
            "mime_type": "application/pdf",
        },
    ])

    # From provider-managed File ID
    response = model.invoke([
        {"type": "text", "text": "Describe the content of this document."},
        {"type": "file", "file_id": "file-abc123"},

>     ])

Reviewer 3: do these need to be `model.invoke(content_blocks=[`?

model_with_tools = model.bind_tools([get_weather])

response = model_with_tools.invoke("What's the weather like in Boston?")
for tool_call in response.tool_calls:
    # View tool calls made by the model
    print(f"Tool: {tool_call['name']}")
    print(f"Args: {tool_call['args']}")
```

When binding user-defined tools, the model's response includes a **request** to
execute a tool. It is up to you to perform the requested action and return the
result back to the model for use in subsequent reasoning.

The LangGraph framework simplifies the orchestration of model and tool calls. See the
[Agent guides](/oss/langchain-agents) to get started. Below, we show some simple
patterns for illustration.

<AccordionGroup>
    <Accordion title="Tool execution loop">
        When a model returns tool calls, you need to execute the tools and pass the
        results back to the model. This creates a conversation loop where the model
        can use tool results to generate its final response.

        ```python Tool execution loop
        # Bind (potentially multiple) tools to the model
        model_with_tools = model.bind_tools([get_weather])

        # Step 1: Model generates tool calls

>         messages = [{"role": "user", "content": "What's the weather in Boston?"}]

Author: why?

Reviewer 2: lighter weight, saves an import.

not long ago we considered using dict format on all input messages throughout docs. langgraph docs for instance almost always use dict format.

don't feel that strongly here.
